lms/services/h_api.py

```diff
@@ -94,20 +94,32 @@ def upsert_user(self, h_user, provider, provider_unique_id):
         except HAPINotFoundError:
             self.create_user(h_user, provider, provider_unique_id)
 
-    def create_group(self, group_id, group_name, creator):
+    def create_group(self, group_id, group_name):
         """
         Create a group in H.
 
         :param group_id: The id of the group
         :param group_name: The display name for the group
-        :param creator: The user creating the group
         """
-        self._api_request(
-            "PUT",
-            f"groups/{group_id}",
-            data={"groupid": group_id, "name": group_name},
-            headers={"X-Forwarded-User": creator.userid},
-        )
+
+        def do_create_group():
+            """Send a create group request to h."""
+            self._api_request(
+                "PUT",
+                f"groups/{group_id}",
+                data={"groupid": group_id, "name": group_name},
+                headers={"X-Forwarded-User": f"acct:lms@{self._authority}"},
+            )
+
+        try:
+            do_create_group()
+        except HAPINotFoundError:
+            # If we get a 404 when trying to create a new group that must mean
+            # that the lms user doesn't exist in h yet.
+            self.create_user(
+                HUser(self._authority, "lms"), provider="lms", provider_unique_id="lms",
+            )
+            do_create_group()
```

Comment:
The create-machine-user logic is specific to h's create-group endpoint. That endpoint 404s if the X-Forwarded-User doesn't exist, so we can detect that situation and create the user. I don't know whether the same is true for other API endpoints. In practice we only use X-Forwarded-User when creating groups. So this logic is implemented here in `create_group()`

Reply:
Unless I'm missing something, doesn't this happen once ever, globally? If that's the case I don't think we should be handling this in every call.

Would this be better as a database migration step to add the user?